First commit for dominant color core code

pbearne opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/56151

pbearne committed:
First commit for dominant color core code

pbearne committed:
debug code for tests

pbearne committed:
debug code for tests

pbearne committed:
debug code for tests

pbearne committed:
debug code for tests

pbearne committed:
debug code for tests

spacedmonkey commented on src/wp-includes/class-wp-image-editor-gd.php:
```diff
@@ -570,4 +570,62 @@ protected function make_image( $filename, $callback, $arguments ) {
+	 *
+	 * @return string|WP_Error Dominant hex color string, or an error on failure.
+	 */
+	public function get_dominant_color() {
```
Need to add these methods to the `WP_Image_Editor` class.

spacedmonkey replied:
These will both need to return WP_Error, with an a message saying these functions are not implemented.

spacedmonkey replied:
Second thoughts maybe this is not the best idea. Let look at https://github.com/WordPress/performance/pull/404

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5449,283 @@ function wp_increase_content_media_count( $amount = 1 ) {
+ * @param int   $attachment_id The attachment ID.
+ * @return array $metadata The attachment metadata.
+ */
+function dominant_color_metadata( $metadata, $attachment_id ) {
```
We can just add this directly there 
https://github.com/WordPress/wordpress-develop/blob/a14324ca3af4eeff69406d3e59719a33f055f07e/src/wp-admin/includes/image.php#L646

pbearne replied:
filter removed and calling function directly

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5449,283 @@ function wp_increase_content_media_count( $amount = 1 ) {
+ * @param object $attachment Image attachment post.
+ * @return mixed $attr Attributes for the image markup.
+ */
+function dominant_color_update_attachment_image_attributes( $attr, $attachment ) {
```
We can add these directly herer - https://github.com/WordPress/wordpress-develop/blob/f52d476bc6dcfe4c5d011f335361500f80708620/src/wp-includes/media.php#L1083

pbearne replied:
filter removed and calling function directly

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5449,283 @@ function wp_increase_content_media_count( $amount = 1 ) {
+ * @param int    $attachment_id  The attachment ID.
+ * @return string image tag
+ */
+function dominant_color_img_tag_add_dominant_color( $filtered_image, $context, $attachment_id ) {
```
Let's not filter and manaully add this in here - https://github.com/WordPress/wordpress-develop/blob/f52d476bc6dcfe4c5d011f335361500f80708620/src/wp-includes/media.php#L1854

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5449,283 @@ function wp_increase_content_media_count( $amount = 1 ) {
+	 *
+	 * You can set this to false in order disable adding the dominant color to the image.
+	 *
+	 * @since 1.2.0
```
```suggestion
	 * @since 6.1
```

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5449,283 @@ function wp_increase_content_media_count( $amount = 1 ) {
+ */
+function _dominant_color_get_dominant_color_data( $attachment_id ) {
+	if ( ! wp_attachment_is_image( $attachment_id ) ) {
+		return new WP_Error( 'no_image_found', __( 'Unable to load image.', 'performance-lab' ) );
```
Remove all the `, 'performance-lab'` references.

pbearne replied:
done

spacedmonkey replied:
Still seeing performance lab

spacedmonkey commented on src/wp-includes/class-wp-image-editor-gd.php:
```diff
@@ -570,4 +570,62 @@ protected function make_image( $filename, $callback, $arguments ) {
+	/**
+	 * Get dominant color from a file.
+	 *
+	 * @since 1.2.0
```
Change all references to 1.2.0 to 6.1

pbearne replied:
done

pbearne committed:
Update src/wp-includes/media.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

pbearne committed:
code review updates

pbearne committed:
Merge remote-tracking branch 'origin/dominant-color-core' into dominant-color-core

pbearne committed:
Merge branch 'WordPress:trunk' into dominant-color-core

pbearne committed:
Merge branch 'WordPress:trunk' into dominant-color-core

pbearne committed:
code review updates

pbearne committed:
renamed functions

pbearne committed:
Fixed old tested to handle dominant color

pbearne committed:
Fixed test with renamed function and removed SVG test

pbearne committed:
whitespace and debug

pbearne committed:
whitespace

pbearne committed:
debug

pbearne committed:
debug

pbearne committed:
debug

pbearne committed:
debug

pbearne committed:
Fixed gif images

pbearne committed:
don't test webp on old version of PHP

pbearne committed:
remove debug

spacedmonkey commented on src/wp-includes/class-wp-image-editor-gd.php:
```diff
@@ -570,4 +570,66 @@ protected function make_image( $filename, $callback, $arguments ) {
+
+		if ( 'image/gif' === $this->mime_type ) {
+			return true;
+		}
```
Let's remove this. And copy what we have in the plugin.

pbearne replied:
done

spacedmonkey commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -951,4 +951,85 @@ protected function pdf_load_source() {
+					return false;
+				}
+			}
+
+			if ( 'image/gif' === $this->image->getImageMimeType() ) {
+				return true;
+			}
```
see https://github.com/WordPress/wordpress-develop/pull/2906/files#r915806634

pbearne replied:
done

pbearne committed:
Fixed test issues

pbearne committed:
test on php versions

pbearne committed:
Removed " so --dominant-color: #949c5f; can added to style

pbearne committed:
Remove webp error test

pbearne committed:
Removed commented code

pbearne committed:
whitespace

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5452,282 @@ function wp_increase_content_media_count( $amount = 1 ) {
+ * @return array|WP_Error Array with the Dominant color and has transparency values or WP_Error on error.
+ */
+function _dominant_color_get_dominant_color_data( $attachment_id ) {
+	if ( ! wp_attachment_is_image( $attachment_id ) ) {
+		return new WP_Error( 'no_image_found', __( 'Unable to load image.' ) );
+	}
```
Update like plugin. 

https://github.com/WordPress/performance/blob/86f53a58f4cf58fa6a083df95aeef46dc44d6ddb/modules/images/dominant-color/load.php#L274-L277

pbearne replied:
done

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5452,282 @@ function wp_increase_content_media_count( $amount = 1 ) {
+ * @return string|null Hex value of dominant color or null if not set.
+ */
+function dominant_color_get_dominant_color( $attachment_id ) {
+	if ( ! wp_attachment_is_image( $attachment_id ) ) {
+		return null;
+	}
```
```suggestion
```

We don't need this anymore.

spacedmonkey commented on tests/phpunit/includes/testcase-dominant-color.php:
```diff
@@ -0,0 +1,140 @@
+				'files_path' => DIR_TESTDATA . '/images/dominant-color/wordpress-gsoc-flyer.pdf',
+			),
+			'mp4' => array(
+				'files_path' => DIR_TESTDATA . '/images/dominant-color/small-video.mp4',
```
Let's use the original 
https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/data/uploads/small-video.mp4

pbearne replied:
swapped and remove file

spacedmonkey commented on tests/phpunit/includes/testcase-dominant-color.php:
```diff
@@ -0,0 +1,140 @@
+	public function provider_get_dominant_color_none_images() {
+		return array(
+			'pdf' => array(
+				'files_path' => DIR_TESTDATA . '/images/dominant-color/wordpress-gsoc-flyer.pdf',
```
Let's reuse this - https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/data/images/wordpress-gsoc-flyer.pdf

pbearne replied:
swapped and remove file

spacedmonkey commented on tests/phpunit/includes/testcase-dominant-color.php:
```diff
@@ -0,0 +1,140 @@
+				'expected_transparency' => true,
+			),
+			'bmp'  => array(
+				'image_path'            => DIR_TESTDATA . '/images/dominant-color/test-image.bmp',
```
Reuse https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/data/images/test-image.bmp

pbearne replied:
swapped and remove file

spacedmonkey commented on tests/phpunit/includes/testcase-dominant-color.php:
```diff
@@ -0,0 +1,140 @@
+	public function provider_get_dominant_color_invalid_images() {
+		$data = array(
+			'tiff' => array(
+				'image_path'            => DIR_TESTDATA . '/images/dominant-color/test-image.tiff',
```
Reuse - https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/data/images/test-image.tiff

pbearne replied:
swapped and remove file

spacedmonkey commented on tests/phpunit/tests/image/dominant-color-image-editor-imageick-test.php:
```diff
@@ -0,0 +1,80 @@
+/**
+ * Tests for dominant-color module.
+ *
+ * @since 1.2.0
```
```suggestion
 * @since 6.1
```

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5452,282 @@ function wp_increase_content_media_count( $amount = 1 ) {
+/**
+ * Get Hex color from RGB.
+ *
+ * @since n.e.x.t
```
```suggestion
 * @since 6.1
```

spacedmonkey commented on tests/phpunit/includes/testcase-dominant-color.php:
```diff
@@ -0,0 +1,140 @@
+				'expected_transparency' => true,
+			),
+		);
+
```
We could add 
https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/data/images/test-image.psd
and maybe even 
https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/data/images/test-image.jp2

pbearne replied:
I can't do jp2 it throws an upload error 
"Sorry, you are not allowed to upload this file type."

spacedmonkey commented on tests/phpunit/tests/image/dominant-color-image-editor-imageick-test.php:
```diff
@@ -0,0 +1,80 @@
+	public $editor_engine = 'WP_Image_Editor_Imagick';
+
+	public function set_up() {
+		parent::set_up(); // TODO: Change the autogenerated stub.
```
Can you explain this todo?

pbearne replied:
no removed

spacedmonkey commented on tests/phpunit/tests/image/dominant-color-image-editor-gd-test.php:
```diff
@@ -0,0 +1,74 @@
+ * @group wp-image-editor-gd
+ * @group dominant-color
+ */
+
```
```suggestion
```

spacedmonkey commented on tests/phpunit/includes/testcase-dominant-color.php:
```diff
@@ -0,0 +1,140 @@
+<?php
+require_once __DIR__ . '/../tests/image/base.php';
+
```
PHP Docs

spacedmonkey commented on tests/phpunit/includes/testcase-dominant-color.php:
```diff
@@ -0,0 +1,140 @@
+<?php
+require_once __DIR__ . '/../tests/image/base.php';
```
Do we need to bootstrap this?

pbearne replied:
yes is error without it
I could move it to the general includes but that feels wrong

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -3894,7 +3898,6 @@ function _wp_image_editor_choose( $args = array() ) {
 
 		if ( isset( $args['methods'] ) &&
 			array_diff( $args['methods'], get_class_methods( $implementation ) ) ) {
-
```
?

pbearne replied:
reverted

spacedmonkey commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -951,4 +951,81 @@ protected function pdf_load_source() {
+			 * has been compiled against ImageMagick version 6.4.0 or newer.
+			 */
+			if ( is_callable( array( $this->image, 'getImageAlphaChannel' ) ) ) {
+				if ( ! $this->image->getImageAlphaChannel() ) {
```
```suggestion
				if ( $this->image->getImageAlphaChannel() === Imagick:: ALPHACHANNEL_UNDEFINED ) {
```

spacedmonkey commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -951,4 +951,81 @@ protected function pdf_load_source() {
+			 * Note that Imagick::getImageAlphaChannel() is only available if Imagick
+			 * has been compiled against ImageMagick version 6.4.0 or newer.
+			 */
+			if ( is_callable( array( $this->image, 'getImageAlphaChannel' ) ) ) {
```
```suggestion
			if ( is_callable( array( $this->image, 'getImageAlphaChannel' ) ) && defined( 'Imagick:: ALPHACHANNEL_UNDEFINED' ) ) {
```

spacedmonkey replied:
See https://github.com/WordPress/wordpress-develop/blob/a14324ca3af4eeff69406d3e59719a33f055f07e/src/wp-includes/class-wp-image-editor-imagick.php#L402-L410

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5452,282 @@ function wp_increase_content_media_count( $amount = 1 ) {
+		$extra_class  = $image_meta['has_transparency'] ? 'has-transparency' : 'not-transparent';
+	}
+
+	if ( ! empty( $data ) || ! empty( $style ) ) {
```
```suggestion
	if ( ! empty( $data ) || ! empty( $style ) || ! str_contains( $style,  $filtered_image )  || ! str_contains( $data,  $filtered_image ) ) {
```

Should we check if this in the string before adding it?

pbearne committed:
Update src/wp-includes/class-wp-image-editor-imagick.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/class-wp-image-editor-imagick.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/media.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/image/dominant-color-image-editor-imageick-test.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

pbearne committed:
Update src/wp-includes/media.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/image/dominant-color-image-editor-gd-test.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

pbearne committed:
Merge branch 'WordPress:trunk' into dominant-color-core

pbearne committed:
changes from code review

pbearne committed:
whirtespaces

pbearne committed:
whirtespaces

pbearne committed:
whirtespaces

spacedmonkey commented:
@pbearne Coding standard issues.

pbearne committed:
Update src/wp-includes/media.php

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

pbearne committed:
Handle old version on Imagick without ALPHACHANNEL_UNDEFINED variable

pbearne committed:
Merge remote-tracking branch 'origin/dominant-color-core' into dominant-color-core

pbearne committed:
Merge branch 'WordPress:trunk' into dominant-color-core

pbearne committed:
whitespace

pbearne committed:
Renamed test files for phpcs

pbearne committed:
debug

pbearne committed:
shortcut gd has_transparency in jpeg

spacedmonkey commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5451,280 @@ function wp_increase_content_media_count( $amount = 1 ) {
+function _dominant_color_get_dominant_color_data( $attachment_id ) {
+	$mime_type = get_post_mime_type( $attachment_id );
+	if ( 'application/pdf' === $mime_type ) {
+		return new WP_Error( 'no_image_found', __( 'Unable to load image.', 'performance-lab' ) );
```
```suggestion
		return new WP_Error( 'no_image_found', __( 'Unable to load image.' ) );
```

spacedmonkey committed:
Apply suggestions from code review

spacedmonkey committed:
Fix unit tests.

spacedmonkey committed:
Check if exists.

pbearne committed:
move the shortcut has_transparency up to calling function

pbearne committed:
whitespace

petitphp commented on src/wp-includes/media.php:
```diff
@@ -5449,3 +5451,285 @@ function wp_increase_content_media_count( $amount = 1 ) {
+ * @param string $size          Optional. Image size. Default 'thumbnail'.
+ * @return false|string Path to an image or false if not found.
+ */
+function wp_get_attachment_file_path( $attachment_id, $size = 'medium' ) {
```
Mismatch between default value for `$size` and the PHPDoc (`medium` vs `thumbnail`)

pbearne replied:
Good spot and thanks
Fixed

pbearne committed:
Update media.php

Fixed a mismatch between default value for $size and the PHPDoc (medium vs thumbnail)

pbearne committed:
Update media.php

reverted none related whitespace edit

pbearne committed:
Update media.php

revert whitespace  edit
